Refactor functoriality and various infrastructure for sequential colimits

VojtechStep opened the pull request:
Builds on #972

VojtechStep commented:
This one is now ready for review

fredrik-bakke commented on src/foundation/binary-homotopies.lagda.md:
```diff
@@ -40,13 +40,21 @@ module _
     (f g : (x : A) (y : B x) → C x y) → UU (l1 ⊔ l2 ⊔ l3)
   binary-htpy f g = (x : A) → f x ~ g x
 
+  binary-htpy' :
+    (f g : (x : A) (y : B x) → C x y) → UU (l1 ⊔ l2 ⊔ l3)
+  binary-htpy' f g = (x : A) → g x ~ f x
```
I've been experimenting with using `inv` or some more descriptive name for definitions such as this one, instead of `'`. For instance, I think a better name for `is-contr-equiv'` is `is-contr-inv-equiv`. Since your definition is not literally inverting homotopies, we could also consider adding a new descriptor like "reverse" or "opposite". In either case, I think the use of `'` here is bad practice. Not that I'm not guilty of using it myself, and at times when there are two competing definitions for the same name and no obvious naming scheme to apply I think `'` is okay to use, but here I don't think that's the case.

VojtechStep replied:
I agree that using `'` to denote an underspecified vibe of "something is the other way around" is bad practice. I'm not opposed to `binary-inv-htpy`, but I also don't have a great argument for calling it that. Since I didn't end up needing this concept, I can just yank it out of this PR, and we can discuss a better naming convention elsewhere; what do you say?

fredrik-bakke replied:
Sounds good!

fredrik-bakke commented on src/synthetic-homotopy-theory/functoriality-sequential-colimits.lagda.md:
````diff
@@ -37,23 +38,38 @@ open import synthetic-homotopy-theory.universal-property-sequential-colimits
+```
+
+The identity morphism is taken to the identity function, `id∞ ~ id`, and
+composition of morphisms is taken to composition of functions,
+`(g ∘ f)∞ ~ g∞ ∘ f∞`.
 
-Furthermore, an
+A corollary of these facts is that taking the
+[standard sequential colimit](synthetic-homotopy-theory.sequential-colimits.md)
+is a functorial action `(A, a) ↦ A∞`.
````
Since this is precisely functoriality, I wouldn't call it a corollary.

VojtechStep replied:
My reasoning for calling it a corollary: what is shown is that given two sequential diagrams with their respective colimits `X` and `Y`, there is an (unnamed) action taking a morphism of sequential diagrams `(A, a) -> (B, b)` to a map `X -> Y`. But this action does not create the cocones on `X` and `Y`, it's merely an action on arrows, so there is no functor to speak of. Only by adding in the mapping of sequential diagrams to their standard sequential colimits, and specializing the preservation properties, do we get a proper map on objects and arrows.

Is it possible that I'm misunderstanding what we call "a functorial action", and it only talks about the action on morphisms?

fredrik-bakke replied:
Sorry, I didn't respond sooner because I don't really understand your rebuttal. Let me try again to understand.

fredrik-bakke replied:
The functor will have type "sequential diagrams of types -> types", that's the type of `standard-sequential-colimit`. I don't think it is, but let me still ask since I'm confused: Is what you're saying that since we don't have the "category of sequential diagrams", there is no functor?

fredrik-bakke replied:
We have so far been quite loose with the term "functoriality" in the library, in part since we don't have a good notion of "category" at our disposal.

VojtechStep replied:
I'll try to make my point clearer.

I agree that with `standard-sequential-colimit`, we have a functor `SD -> Types`, where SD is the category of sequential diagrams.

However, the first half of the file only speaks about a map of morphisms, which requires more parameters than just a morphism of sequential diagrams: "given diagrams A, B, and their sequential colimits X, Y, we have a map `SD(A, B) -> Types(X, Y)`". At that point, we don't yet have a functor `SD -> Types`, because we don't have a map on objects; the `X` and `Y` are taken as inputs to this action.

fredrik-bakke replied:
Okay, I hope I understand you better now. The objects exist by the existence of sequential colimts, they are always essentially unique by the universal property, and they are unique given univalence. To me, it seems like you're complicating the topic by not just calling it functoriality. But if you insist on keeping it, perhaps the prose reads better if you replace "as a corollary" with "in particular" or "as a consequence". This blurs the distinction at least a little bit.

VojtechStep replied:
I reworded that section a bit

fredrik-bakke replied:
There is another good reason for your formalization though. It gives a more implementation-agnostic treatment of the topic and avoids the unnecessary use of univalence, thus giving better computational behavior.

fredrik-bakke commented on src/synthetic-homotopy-theory/functoriality-sequential-colimits.lagda.md:
```diff
@@ -37,23 +38,38 @@ open import synthetic-homotopy-theory.universal-property-sequential-colimits
+induces an [equivalence](foundation.equivalences.md) of their colimits.
 
-The development in this file is a formalization of Lemma 3.5 in Sequential
+The development in this file is a generalization of Lemma 3.5 from Sequential
```
Perhaps it is best at this point to leave it at referencing the article in the `## References` section, since the file's contents may change irrespective of what is in that article.

fredrik-bakke commented on src/synthetic-homotopy-theory/functoriality-sequential-colimits.lagda.md:
````diff
@@ -72,24 +88,24 @@ module _
+    A₀ ---> A₁ ---> A₂ ---> ⋯ ---> X
     |       |       |              |
  f₀ |       | f₁    | f₂           | f∞
     V       V       V              V
-    B₀ ---> B₁ ---> B₂ ---> ⋯ ---> B∞
+    B₀ ---> B₁ ---> B₂ ---> ⋯ ---> Y
         b₀      b₁      b₂
 ```
 
 then induces a
 [cocone](synthetic-homotopy-theory.cocones-under-sequential-diagrams.md) under
-`(A, a)` with codomain `B∞`, which is homotopic to the standard cocone under
-`(B, b)` precomposed by `f`.
+`(A, a)` with codomain `Y`, which is homotopic to the cocone under `(B, b)`
````
There is nothing wrong with writing `A∞` for any (non-standard) colimit of `A` in my opinion. Alternatively, you may want to consider giving `f∞` a different name, if you want to reserve that notation for the standard colimit.

VojtechStep replied:
Denoting the arbitrary colimits `<diagram>∞` makes it seem a bit too much like they are outputs of the theorems, rather than inputs IMO. On the other hand, I do think of `f∞` as the "standard sequential colimit" of the sequence of maps in the morphism, constructed in the theorems, even when `X` and `Y` are non-standard, which is why I added the `∞` mark. Do you object to this line of thought, or have a better convention in mind?

EgbertRijke replied:
Probably it would be reasonable to call this map `g`. The `f`'s go between `A`'s and `B`'s, but `X` and `Y` aren't `A`'s and `B`'s so it would make sense to me to give it a new name `g`.

VojtechStep replied:
I changed it to `g`.

fredrik-bakke commented on src/synthetic-homotopy-theory/functoriality-sequential-colimits.lagda.md:
```diff
@@ -99,21 +115,431 @@ This homotopy of cocones provides
+  { f g : hom-sequential-diagram A B} (H : htpy-hom-sequential-diagram B f g)
+  where
+
+  htpy-map-sequential-colimit-htpy-hom-sequential-diagram :
+    map-sequential-colimit-hom-sequential-diagram up-c c' f ~
+    map-sequential-colimit-hom-sequential-diagram up-c c' g
+  htpy-map-sequential-colimit-htpy-hom-sequential-diagram =
+    htpy-eq
+      ( ap
+        ( map-sequential-colimit-hom-sequential-diagram up-c c')
+        ( eq-htpy-sequential-diagram _ _ f g H))
```
In case you didn't consider it, I tend to look for ways to define constructions such as this one without applying function extensionality. This gives definitions with nicer computational behavior down the line.

VojtechStep replied:
I did consider it, but currently it's only used in abstract definitions, so I preferred a short "opaque" definition over an explicit computable one. If it ever becomes desirable to have a more elaborate definition, we can come back to it, like Raymond did with #924 for suspensions.

fredrik-bakke commented on src/synthetic-homotopy-theory/functoriality-sequential-colimits.lagda.md:
```diff
@@ -99,21 +115,431 @@ This homotopy of cocones provides
+      ( ( map-sequential-colimit-hom-sequential-diagram up-c' c'' g) ∘
+        ( map-sequential-colimit-hom-sequential-diagram up-c c' f))
+  pr1 htpy-preserves-comp-map-sequential-colimit-hom-sequential-diagram n =
+    ( htpy-htpy-cocone-map-sequential-colimit-hom-sequential-diagram up-c c''
+      ( comp-hom-sequential-diagram A B C g f)
+      ( n)) ∙h
+    pasting-vertical-coherence-square-maps
```
Inconsistent parenthization. The first argument of `_∙h_` is parenthesized while the second is not.

VojtechStep replied:
Good catch!

fredrik-bakke commented on src/synthetic-homotopy-theory/functoriality-sequential-colimits.lagda.md:
By the way! I think we should form a convention regarding whether we prefer formulating frameworks such as this one in terms of the small proposition or the large universal property of a (co)limit. For instance, most results regarding pullbacks are phrased in terms of `is-pullback` and not `universal-property-pullback`. The small property has the advantage that it is small so can, for instance, be used in Sigma types, and can leverage the explicit description of the standard (co)limit, while the large universal property has the advantage that it makes sense irrespective of the existence of the (co)limit, and is more "direct".

fredrik-bakke replied:
It doesn't make sense to me to repeat every result for every possible equivalent definition, so I think forming a convention could save us some work.

VojtechStep replied:
I personally prefer the universal properties, but I don't have a great basis to do so - I'm not very fond of having the `is-thing` properties reference the standard constructions, since 1) eventually they don't need to exist, and 2) there is an underlying direction for the equivalence to the standard construction.

If we felt the need to bundle a (co)cone with a universal property, we could build some infrastructure for large(?) subtypes, which are given by a family of large propositions over a type.

EgbertRijke replied:
I am siding with Fredrik on this one. Some isntances where we definitely prefer the small proposition of being a pullback of pushout is in the definitions of (co)cartesian morphisms of arrows. I could see a similar definition for sequential composites of maps being useful at some point.

I don't wish that we develop a notion of large propositions (or a foundation for large types, for that matter). Especially for reasons such as above, which are easily avoided by having small `is-colimit` predicates.

EgbertRijke replied:
Fredrik has added at some point large sigma types and large identity types to the library, and at several points in time I have been very tempted to delete these files:)

fredrik-bakke replied:
I'm not opposed to having them removed either

fredrik-bakke commented on src/synthetic-homotopy-theory/functoriality-sequential-colimits.lagda.md:
````diff
@@ -99,21 +115,431 @@ This homotopy of cocones provides
+    is-equiv-map-equiv-sequential-colimit-equiv-sequential-diagram
+```
+
+### A morphism of sequential diagrams induces a map of standard sequential colimits
````
There's a couple of ways to organize this file. Either, you can

1. group all results regarding retracts together, in terms of universal property and standard colimit and so on,

or you can

2. group all results regarding universal property together, and then standard colimits.

If you want to continue with the second approach, I think it warrants a header separating the two sections. I tend to organize my files following the first approach.

VojtechStep replied:
I stuck to the second approach, as I consider the whole section with the standard sequential colimits an application of the results above, rather than part of the backing theory.

fredrik-bakke reviewed (approved):
Nice work! See my comments 😊

EgbertRijke commented on src/synthetic-homotopy-theory/cocones-under-sequential-diagrams.lagda.md:
````diff
@@ -76,7 +76,7 @@ module _
 ```agda
 module _
-  { l1 l2 : Level} (A : sequential-diagram l1) {X : UU l2}
+  { l1 l2 : Level} {A : sequential-diagram l1} {X : UU l2}
````
I would be surprised if Agda can always figure out the sequential diagram as an implicit argument.

VojtechStep replied:
There are 46 occurrences of `map-cocone-sequential-diagram` being used, and not a single one fails to infer, so I made it implicit. We can revert it to being explicit if it causes problems down the line, but for the time being it works well.

EgbertRijke commented on src/synthetic-homotopy-theory/functoriality-sequential-colimits.lagda.md:
```diff
@@ -37,24 +38,44 @@ open import synthetic-homotopy-theory.universal-property-sequential-colimits
+[sequential colimits](synthetic-homotopy-theory.universal-property-sequential-colimits.md)
+`X` and `Y`, and a
+[morphism](synthetic-homotopy-theory.morphisms-sequential-diagrams.md)
+`f : (A, a) → (B, b)`, we get an induced map `f∞ : X → Y`, as in the following
+diagram:
```
...there is a unique map `g : X → Y` such that the diagram ... commutes

EgbertRijke commented on src/synthetic-homotopy-theory/functoriality-sequential-colimits.lagda.md:
```diff
@@ -72,24 +93,24 @@ module _
-### A morphism of sequential diagrams induces a map of standard sequential colimits
+### A morphism of sequential diagrams induces a map of sequential colimits
 
 The induced map
```
The unique map `g : X -> Y` such that the diagram ... commutes then induces...

EgbertRijke commented on src/synthetic-homotopy-theory/functoriality-sequential-colimits.lagda.md:
```diff
@@ -99,285 +120,581 @@ This homotopy of cocones provides
+    Aₙ ---------> X
     |      |      |
     |      V      |
  fₙ |     Bₙ₊₁    | f∞
```
g

EgbertRijke commented on src/synthetic-homotopy-theory/functoriality-sequential-colimits.lagda.md:
```diff
@@ -99,285 +120,581 @@ This homotopy of cocones provides
-coherences of the cocones of the standard sequential colimits, the back left
+coherences of the cocones of the sequential colimits, the back left
 [square](foundation-core.commuting-triangles-of-maps.md) is coherence of `f`,
 and the front and back right squares are provided by uniqueness of `f∞`.
```
...by the uniqueness of `g`.

EgbertRijke commented:
Would it be possible to work towards merging this PR?

VojtechStep commented:
Yes, I'll be at my computer in like an hour, and I'll try to address the comments

VojtechStep commented:
I rebased on master and addressed most of the suggestions. The last thing left open is my usage of the term "corollary", which I explained, and I'm waiting for a reaction

fredrik-bakke commented on src/synthetic-homotopy-theory/cocones-under-sequential-diagrams.lagda.md:
```diff
@@ -201,53 +200,53 @@ captured by a homotopy between them.
   reflexive-htpy-cocone-sequential-diagram :
-    ( c : cocone-sequential-diagram A X) → htpy-cocone-sequential-diagram A c c
-  pr1 (reflexive-htpy-cocone-sequential-diagram c) n = refl-htpy
-  pr2 (reflexive-htpy-cocone-sequential-diagram c) n = right-unit-htpy
+    htpy-cocone-sequential-diagram c c
+  pr1 reflexive-htpy-cocone-sequential-diagram n = refl-htpy
+  pr2 reflexive-htpy-cocone-sequential-diagram n = right-unit-htpy
```
For consistency, I suggest using `refl`. We have 363 definitions using this string for a proof of reflexivity, while only 10 use `reflexive`.

fredrik-bakke replied:
Sorry for not catching this when it was introduced btw

VojtechStep replied:
When writing this I remembered there were some refactors changing one to the other, so I checked cocones under spans for a reference, expecting that to conform to the current conventions. Well, apparently that's the only `reflexive-` one left in synthetic homotopy theory (apart from `26-...` stuff) 😅.

It should be fixed now.

EgbertRijke commented:
I'm happy with the changes! I think I'm just gonna go ahead and merge this one

EgbertRijke commented:
Thank you for this PR Volta!
